### PR TITLE
Add ReplicaCreationService

### PR DESCRIPTION
**Pull request by bryanlb**
Creates replicas when changes in snapshots are detected.

### CODE REVIEW

File: kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    meterRegistry = new SimpleMeterRegistry();
+    testingServer = new TestingServer();
+  }
+
+  @After
+  public void shutdown() throws IOException {
+    testingServer.close();
+    meterRegistry.close();
+  }
+
```
Comment (mansu) [inline kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java:45]: We should also add unit tests when stores are empty and when first snapshot is added. 

Also, we should also add some unit tests to make sure that this code is robust against ZK failures. For example, if we fail to create a replica, the next run of the program should create it.

File: kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    assertThat(
+            (int)
+                replicaMetadataStore
+                    .getCached()
+                    .stream()
+                    .filter(replicaMetadata -> Objects.equals(replicaMetadata.snapshotId, "b"))
+                    .count())
+        .isEqualTo(REPLICAS_PER_SNAPSHOT_TEST);
+
+    replicaMetadataStore.close();
```
Comment (mansu) [inline kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java:158]: Can we move the close operations into tearDown?

File: kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                MetricsUtil.getCount(ReplicaCreatorService.REPLICAS_CREATED, meterRegistry)
+                    == REPLICAS_PER_SNAPSHOT_TEST * 2);
+
+    ReplicaMetadataStore replicaMetadataStore =
+        new ReplicaMetadataStore(metadataStoreService.getMetadataStore(), true);
+    assertThat(replicaMetadataStore.getCached().size()).isEqualTo(4);
+    assertThat(
+            (int)
+                replicaMetadataStore
+                    .getCached()
```
Comment (mansu) [inline kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java:144]: In unit tests prefer listSync calls to check metadata instead of cached calls which may be lagging due to timing issues.

I think it would be a good idea to add a check to make sure that the existing data before the assignment is still the same after the assignment.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java [hunk 4]
```diff
@@ -39,23 +48,46 @@ protected void startUp() throws Exception {
...
   }
 
-  private KaldbMetadataStoreChangeListener replicaNodeListener() {
-    return () -> {
-      List<ReplicaMetadata> replicaMetadataList = replicaMetadataStore.getCached();
-      LOG.debug(
-          "Change on replica metadata, new replica metadata count {}", replicaMetadataList.size());
-    };
+  private void createReplicasForUnassignedSnapshots() {
+    snapshotMetadataStore
```
Comment (mansu) [inline kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java:72]: Add logging before and after this process runs. 

Add a counter so we know how many times this process has run. 

Also, it may be better to add a metrics to measure how long a single execution of this process took to run.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java [hunk 5]
```diff
@@ -39,23 +48,46 @@ protected void startUp() throws Exception {
...
+    // handled
+    createReplicasForUnassignedSnapshots();
   }
 
   @Override
   protected void shutDown() throws Exception {
     LOG.info("Closing replica create service");
 
     replicaMetadataStore.close();
+    snapshotMetadataStore.close();
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java:66]: Since we're moving these stores up a level and passing them in, it won't be necessary to do anything special here. We also don't need to un-register listeners explicitly, as this is already handled as as part of the abstract base class. 

https://github.com/slackhq/kaldb/blob/master/kaldb/src/main/java/com/slack/kaldb/metadata/core/CacheableMetadataStore.java#L81

File: kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java (hunk 3 above)
Comment (bryanlb) [reply kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java:144]: As the underlying implementation is written against the cached view, when the test runs it should be using the same underlying data structure. 

I can add an an extra `await()` here however. Technically in line 132 we've already guaranteed that the service has bootstrapped and initialized all the data structures, but it would be possible under heavy load that the cache view hasn't been updated yet.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java [hunk 6]
```diff
@@ -22,15 +23,23 @@
...
 
   private ReplicaMetadataStore replicaMetadataStore;
+  private SnapshotMetadataStore snapshotMetadataStore;
+
+  public static final String REPLICAS_CREATED = "replicas_created";
+  private final Counter replicasCreated;
 
   public ReplicaCreatorService(
-      MetadataStoreService metadataStoreService, MeterRegistry meterRegistry) {
+      MetadataStoreService metadataStoreService,
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java]: Done.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java [hunk 7]
```diff
@@ -39,23 +48,46 @@ protected void startUp() throws Exception {
...
-      LOG.debug(
-          "Change on replica metadata, new replica metadata count {}", replicaMetadataList.size());
-    };
+  private void createReplicasForUnassignedSnapshots() {
+    snapshotMetadataStore
+        .getCached()
+        .stream()
+        .filter(
+            snapshotMetadata ->
+                replicaMetadataStore
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java:70]: Refactored to create a single `HashSet` before operating on the snapshot list. This guarantees that we have a constant-time check against `contains()`, and is also a stable set.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java [hunk 8]
```diff
@@ -39,23 +48,46 @@ protected void startUp() throws Exception {
...
     replicaMetadataStore = new ReplicaMetadataStore(metadataStoreService.getMetadataStore(), true);
-    replicaMetadataStore.addListener(replicaNodeListener());
+
+    snapshotMetadataStore =
+        new SnapshotMetadataStore(metadataStoreService.getMetadataStore(), true);
+    snapshotMetadataStore.addListener(this::createReplicasForUnassignedSnapshots);
+
+    // we need to fire this at startup, so any snapshots that were created while we were offline are
+    // handled
+    createReplicasForUnassignedSnapshots();
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java:52]: Done.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java (hunk 4 above)
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java:77]: Done. Switched to a timer so that we get both the count and the time information.

File: kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java (hunk 2 above)
Comment (bryanlb) [reply kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java:161]: Done.

File: kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    SnapshotMetadata snapshotB =
+        new SnapshotMetadata(
+            "b", "b", Instant.now().toEpochMilli() - 1, Instant.now().toEpochMilli(), 0, "b");
+
+    snapshotMetadataStore.createSync(snapshotA);
+    snapshotMetadataStore.createSync(snapshotB);
+
+    // create one replica for A, two for B
+    replicaMetadataStore.createSync(
+        new ReplicaMetadata(UUID.randomUUID().toString(), snapshotA.snapshotId));
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java:93]: Done.

File: kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java (hunk 3 above)
Comment (bryanlb) [reply kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java:136]: Per our discussion I went ahead and added an extra check to `listSync`, as well as the `await()` on the `getCached` call.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java [hunk 10]
```diff
@@ -39,23 +48,46 @@ protected void startUp() throws Exception {
...
+        .stream()
+        .filter(
+            snapshotMetadata ->
+                replicaMetadataStore
+                    .getCached()
+                    .stream()
+                    .noneMatch(
+                        replicaMetadata ->
+                            replicaMetadata.snapshotId.equals(snapshotMetadata.snapshotId)))
+        .forEach(
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java:93]: Added better detection of replica creation failures by adding logging and metrics for when this happens.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java [hunk 11]
```diff
@@ -39,23 +48,46 @@ protected void startUp() throws Exception {
     metadataStoreService.awaitRunning(DEFAULT_START_STOP_DURATION);
 
     replicaMetadataStore = new ReplicaMetadataStore(metadataStoreService.getMetadataStore(), true);
-    replicaMetadataStore.addListener(replicaNodeListener());
+
+    snapshotMetadataStore =
+        new SnapshotMetadataStore(metadataStoreService.getMetadataStore(), true);
+    snapshotMetadataStore.addListener(this::createReplicasForUnassignedSnapshots);
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreatorService.java]: Done. Added an event aggregation/debounce window, and a scheduled service. Also exposed the appropriate params into the cluster manager configs as a separate nested section.

File: kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java (hunk 1 above)
Comment (bryanlb) [reply kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java]: Done.

File: kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    replicaMetadataStore.createSync(
+        new ReplicaMetadata(UUID.randomUUID().toString(), snapshotB.snapshotId));
+    replicaMetadataStore.createSync(
+        new ReplicaMetadata(UUID.randomUUID().toString(), snapshotB.snapshotId));
+
+    ReplicaCreatorService replicaCreatorService =
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java]: Done.

File: kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreationServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,492 @@
...
+        .isEqualTo(2);
+    assertThat(MetricsUtil.getCount(ReplicaCreationService.REPLICAS_FAILED, meterRegistry))
+        .isZero();
+
+    replicaCreationService.stopAsync();
+    replicaCreationService.awaitTerminated(DEFAULT_START_STOP_DURATION);
+  }
+
+  @Test
+  public void shouldStillCreateReplicaIfFirstAttemptFails() throws Exception {
```
Comment (mansu) [inline kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreationServiceTest.java:298]: Another way to test this would be to succeed in creating replicas once, then delete a replica and then run the assignment again to see if it we come back with a full list of replicas.

File: kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreationServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,492 @@
...
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.IntStream;
+import org.apache.curator.test.TestingServer;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ReplicaCreationServiceTest {
```
Comment (mansu) [inline kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreationServiceTest.java:42]: I think a unit test ensuring that we throw an exception on bad config values would be very valuable here.

**Review by mansu — CHANGES_REQUESTED** [review]

**Comment by mansu** [comment]
Thanks for making the changes. The code looks great! Just a few minor updates and this should be good to go.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreationService.java [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+  /**
+   * Creates N replicas per the KalDb configuration for each snapshot. If a snapshot does not
+   * contain the amount of replicas configured this will attempt to create the missing replicas to
+   * bring it into compliance.
+   *
+   * @return The count of successful created replicas
+   */
+  protected int createReplicasForUnassignedSnapshots() {
+    LOG.info("Starting replica creation for unassigned snapshots");
+    Timer.Sample assignmentTimer = Timer.start(meterRegistry);
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreationService.java:130]: It does - when using micometer with prometheus you get both [the count and total time](https://micrometer.io/docs/registry/prometheus#_timers), which can be used to also visualize the average latency and throughput.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreationService.java [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+                        .mapToObj(
+                            (i) ->
+                                replicaMetadataStore.create(
+                                    replicaMetadataFromSnapshotId(snapshotMetadata.snapshotId)))
+                        .collect(Collectors.toList()))
+            .flatMap(List::stream)
+            .collect(Collectors.toList());
+
+    ListenableFuture<List<Object>> futureList = Futures.allAsList(createdReplicaMetadataList);
+    int completeFutures = 0;
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreationService.java]: I went with `createdReplicas` and `failedReplicas` as this better matches how we're using these. I decided I didn't like referring to these counters as `completeFutures/incompleteFutures` as the definition of what a complete future would be wasn't clear (a future that was "done" wasn't necessarily "completed").

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreationService.java [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+  protected Scheduler scheduler() {
+    // run one iteration getScheduleInitialDelayMins after startup, and then every
+    // getSchedulePeriodMins mins after that
+    return Scheduler.newFixedRateSchedule(
+        replicaCreationServiceConfig.getScheduleInitialDelayMins(),
+        replicaCreationServiceConfig.getSchedulePeriodMins(),
+        TimeUnit.MINUTES);
+  }
+
+  /**
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreationService.java:121]: Done.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreationService.java [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+
+    ListenableFuture<List<Object>> futureList = Futures.allAsList(createdReplicaMetadataList);
+    int completeFutures = 0;
+    int incompleteFutures = 0;
+
+    try {
+      completeFutures = futureList.get(DEFAULT_ZK_TIMEOUT_SECS, TimeUnit.SECONDS).size();
+    } catch (Exception futureListException) {
+      for (ListenableFuture<?> future : createdReplicaMetadataList) {
+        if (future.isDone()) {
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreationService.java:177]: Done.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreationService.java [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+        "Completed replica creation for unassigned snapshots - successfully created {} replicas, failed {} replicas in {} ms",
+        completeFutures,
+        incompleteFutures,
+        TimeUnit.MILLISECONDS.convert(assignmentDuration, TimeUnit.NANOSECONDS));
+
+    return completeFutures;
+  }
+
+  public static ReplicaMetadata replicaMetadataFromSnapshotId(String snapshotId) {
+    return new ReplicaMetadata(UUID.randomUUID().toString(), snapshotId);
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreationService.java:208]: Added a created time, and changed the replica name to follow the pattern of `snapshotId-UUID`

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreationService.java [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+  public ReplicaCreationService(
+      ReplicaMetadataStore replicaMetadataStore,
+      SnapshotMetadataStore snapshotMetadataStore,
+      KaldbConfigs.ManagerConfig.ReplicaCreationServiceConfig replicaCreationServiceConfig,
+      KaldbConfigs.ManagerConfig.ReplicaEvictionServiceConfig replicaEvictionServiceConfig,
+      MeterRegistry meterRegistry) {
+
+    this.replicaMetadataStore = replicaMetadataStore;
+    this.snapshotMetadataStore = snapshotMetadataStore;
+    this.replicaCreationServiceConfig = replicaCreationServiceConfig;
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreationService.java:68]: This isn't necessary, as these checks are already done as part of the `AbstractScheduledService` base class.

[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaCreationService.java:68]: Added preconditions to ensure that the ones we are responsible for are reasonable values. The schedule configs are already checked as part of the `AbstractScheduledService` so added a comment noting that.

File: kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreationServiceTest.java (hunk 14 above)
Comment (bryanlb) [reply kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreationServiceTest.java:42]: Done, tests added for each of our precondition checks.

File: kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreationServiceTest.java [hunk 21]
```diff
@@ -0,0 +1,492 @@
...
+            replicaEvictionServiceConfig,
+            meterRegistry);
+    replicaCreationService.startAsync();
+    replicaCreationService.awaitRunning(DEFAULT_START_STOP_DURATION);
+
+    assertThat(MetricsUtil.getCount(ReplicaCreationService.REPLICAS_CREATED, meterRegistry))
+        .isEqualTo(0);
+
+    replicaCreationService.stopAsync();
+    replicaCreationService.awaitTerminated(DEFAULT_START_STOP_DURATION);
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreationServiceTest.java:122]: Done.

File: kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreationServiceTest.java [hunk 22]
```diff
@@ -0,0 +1,492 @@
...
+  public void shouldCreateReplicasIfNoneExist() throws Exception {
+    SnapshotMetadata snapshotA =
+        new SnapshotMetadata(
+            "a", "a", Instant.now().toEpochMilli() - 1, Instant.now().toEpochMilli(), 0, "a");
+    snapshotMetadataStore.createSync(snapshotA);
+
+    KaldbConfigs.ManagerConfig.ReplicaCreationServiceConfig replicaCreationServiceConfig =
+        KaldbConfigs.ManagerConfig.ReplicaCreationServiceConfig.newBuilder()
+            .setEventAggregationSecs(2)
+            .setReplicasPerSnapshot(4)
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreationServiceTest.java:135]: I added a separate test suite for this to test that the 0 case is handled as expected.

**Review by mansu — APPROVED** [review]
Thanks for the changes.
